### PR TITLE
RtD docs previews: Cancel building PRs if no changes in Doc dir

### PR DESCRIPTION
**Pull request by hugovk**
As suggested by @humitos at https://github.com/python/cpython/pull/103843#discussion_r1181870475, cancel building pull requests when there aren't changes in the `Doc` directory.

If there are no changes (`git diff` exits with 0) we force the command to return with 183. This is a special exit code on Read the Docs that will cancel the build immediately.

https://docs.readthedocs.io/en/stable/build-customization.html#cancel-build-based-on-a-condition

### CODE REVIEW

**Comment by hugovk** [comment]
For the first commit:

> Problem in your project's configuration. Invalid "build.commands": .readthedocs.yml: The keys build.jobs and build.commands can't be used together.

https://readthedocs.org/projects/cpython-previews/builds/20362322/

For the second:

> We encountered a problem with a command while building your project. To resolve this error, double check your project configuration and installed dependencies are correct and have not changed recently.
> ...
> /bin/sh: 1: if: not found
> /bin/sh: 2: Syntax error: "then" unexpected

https://readthedocs.org/projects/cpython-previews/builds/20362458/

@humitos Any suggestions? Thank you!

**Comment by Eclips4** [comment]
Just guessing, maybe move all commands and lines of bash code into jobs/post_checkout section will help?

File: .readthedocs.yml [hunk 1]
```diff
@@ -13,6 +13,19 @@ build:
...
   commands:
+    # https://docs.readthedocs.io/en/stable/build-customization.html#cancel-build-based-on-a-condition
+    #
+    # Cancel building pull requests when there aren't changes in the Doc directory.
+    #
+    # If there are no changes (git diff exits with 0) we force the command to return with 183.
+    # This is a special exit code on Read the Docs that will cancel the build immediately.
+    - |
+      if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && git diff --quiet origin/main -- Doc/ .readthedocs.yml;
+      then
```
Comment (sobolevn) [inline .readthedocs.yml]:
```suggestion
      if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && git diff --quiet origin/main -- Doc/ .readthedocs.yml; then
```

Comment (hugovk) [reply .readthedocs.yml]: Hmm, similar error:

```
if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && git diff --quiet origin/main -- Doc/ .readthedocs.yml; then
  echo "No changes to Doc/ - exiting the build.";
  exit 183;
fi
/bin/sh: 1: Syntax error: "then" unexpected
```

**Comment by sobolevn** [comment]
Hm, looks like `sh` cannot handle `] && expr` case. Better use `[ ]` twice.

File: .readthedocs.yml [hunk 2]
```diff
@@ -13,6 +13,18 @@ build:
...
 
   commands:
+    # https://docs.readthedocs.io/en/stable/build-customization.html#cancel-build-based-on-a-condition
+    #
+    # Cancel building pull requests when there aren't changes in the Doc directory.
+    #
+    # If there are no changes (git diff exits with 0) we force the command to return with 183.
+    # This is a special exit code on Read the Docs that will cancel the build immediately.
+    - |
+      if [[ "$READTHEDOCS_VERSION_TYPE" = "external" ]] && git diff --quiet origin/main -- Doc/ .readthedocs.yml; then
```
Comment (sobolevn) [inline .readthedocs.yml]:
```suggestion
      if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ]; then
```

**Review by sobolevn — COMMENTED** [review]
Sorry, I should have posted the full version right away :)

**Comment by humitos** [comment]
Hi! Sorry for the late reply. I've been on vacations and then Write the Docs conference after that.

I was 100% sure that we already solved the problem you are facing, but I just found that we reverted the change because it was causing other issues (https://github.com/readthedocs/readthedocs.org/pull/10206). I'm sorry for the confusion here.

As @Eclips4 commented, putting this into a script and calling it from `build.jobs.post_` should work as a workaround.

**Comment by humitos** [comment]
Interesting... It seems it works fine with `build.jobs` (I tested it again at https://readthedocs.org/projects/test-builds/builds/20716439/) but it doesn't with `build.commands` for some reason. I will need to keep researching about this and I will come back, hopefully, with a solution 😄

**Comment by sobolevn** [comment]
It works locally, maybe this is some kind of `yaml` issue?

```sh
sh-3.2$ if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ]; then
>   echo 1
> fi
sh-3.2$ 
```

**Comment by sobolevn** [comment]
Maybe like this?

```yaml
- if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ]; then \
    echo "No changes to Doc/ - exiting the build."; \
    exit 183; \
  fi
```

**Comment by sobolevn** [comment]
I have no idea what is going on :)
Maybe we are missing something obvious?

**Comment by humitos** [comment]
Today we are deploying a change on Read the Docs that should fix this. Hopefully, it works 🤞🏼

**Comment by humitos** [comment]
The fix on Read the Docs is already deployed and my tests worked. Can you confirm that it works for you now as well? Note that you have to adapt your command to revert it to the state it was before (multi-line). I will do a suggestion on the PR.

File: .readthedocs.yml [hunk 3]
```diff
@@ -13,6 +13,18 @@ build:
...
+    # https://docs.readthedocs.io/en/stable/build-customization.html#cancel-build-based-on-a-condition
+    #
+    # Cancel building pull requests when there aren't changes in the Doc directory.
+    #
+    # If there are no changes (git diff exits with 0) we force the command to return with 183.
+    # This is a special exit code on Read the Docs that will cancel the build immediately.
+    - if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ]; then \
+        echo "No changes to Doc/ - exiting the build."; \
+        exit 183; \
+      fi
```
Comment (humitos) [inline .readthedocs.yml]:
```suggestion
    - |
      if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ];
      then
        echo "No changes to Doc/ - exiting the build.";
        exit 183;
    fi
```

I'm not 100% the indentation is correct, but the idea here is to use `|` in YAML to define it as multi-line and remove all the trailing `\` from the command.

I think it should work this way now that we deployed a fix on RTD.

Comment (humitos) [reply .readthedocs.yml]: BTW, in my test it works with the trailing `\`. It seems that I only added the `|` at the beginning, but I suppose removing the `\` should also work: https://readthedocs.org/projects/test-builds/builds/20793631/

**Comment by arhadthedev** [comment]
Merging `main` to trigger the documentation build.

File: .readthedocs.yml (hunk 3 above)
Comment (hugovk) [reply .readthedocs.yml]: Looks like it's working now, thanks!

Here's more about YAML multilines... https://yaml-multiline.info

I think it's neater like this with `|` and no trailing `\`s.

**Comment by hugovk** [comment]
This is working quite well, thanks!

One thing though, for example on https://github.com/python/cpython/pull/105008#event-9355526987, the build was cancelled as expected:

<details>
<summary>Details</summary>

image


</details>

https://readthedocs.org/projects/cpython-previews/builds/20835824/

However, the status check on GitHub still shows as pending:

<details>
<summary>Details</summary>

image


</details>

And also shows in the tab:

<details>
<summary>Details</summary>

image


</details>

Would it be possible to have it report as skipped instead of pending?

**Comment by humitos** [comment]
Hi @hugovk. This should be reporting success (✅) to GitHub when the build is skipped. It seems we introduced a bug here or something that broke this behavior. If you have the time, please open an issue in https://github.com/readthedocs/readthedocs.org/

**Comment by hugovk** [comment]
Thanks, reported: https://github.com/readthedocs/readthedocs.org/issues/10364

---

Another PR had no `Doc/` or `.readthedocs.yml` changes but still built the docs preview:

* https://github.com/python/cpython/pull/105058
* https://github.com/python/cpython/pull/105058/commits/de000c6fe7af2a9eb6019813244752f546d6b0cb
* https://readthedocs.org/projects/cpython-previews/builds/20850890/
* https://cpython-previews--105058.org.readthedocs.build/en/105058/

We can see the check was there but returned 0:

```sh
if [ "$READTHEDOCS_VERSION_TYPE" = "external" ] && [ "$(git diff --quiet origin/main -- Doc/ .readthedocs.yml; echo $?)" -eq 0 ];
then
  echo "No changes to Doc/ - exiting the build.";
  exit 183;
fi
```

Any ideas why this didn't skip?

**Pull request merged**